### PR TITLE
Aditya's PR 18

### PR DESCRIPTION
**Pull request by potuz**
This PR implements PR18. It also adds a prunning mechanism to remove any descendant of finalized checkpoint that is incompatible with its finalization.

### CODE REVIEW

File: beacon-chain/forkchoice/doubly-linked-tree/store.go [hunk 1]
```diff
@@ -238,6 +238,22 @@ func (s *Store) prune(ctx context.Context) error {
...
+		return nil
+	}
+
+	for _, child := range finalizedNode.children {
+		if child != nil && child.slot <= checkpointMaxSlot {
+			if err := s.pruneFinalizedNodeByRootMap(ctx, child, finalizedNode); err != nil {
+				return errors.Wrap(err, "could not prune incompatible finalized child")
+			}
+		}
+	}
```
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:256]: Is it ok that this is not gated under feature flag?

Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/store.go:256]: yeah this is fine, this removes all nodes that are incompatible with finalization, we never want to continue building on those.

File: beacon-chain/state/stategen/getter.go [hunk 2]
```diff
@@ -65,7 +65,7 @@ func (s *State) StateByRoot(ctx context.Context, blockRoot [32]byte) (state.Beac
 	return s.loadStateByRoot(ctx, blockRoot)
 }
 
-// BalancesByRoot retrieves the effective balances of all validators at the
+// BalancesByRoot retrieves the effective balances of all active validators at the
```
Comment (terencechain) [inline beacon-chain/state/stategen/getter.go:68]: Can you please update the function name and the comments to indicate these is non-slashed balances? 
While it's ok now, I'm afraid someone tomorrow decide to use this function for consensus-related functions and other clients still account for slashed balances

**Review by terencechain — APPROVED** [review]
Lgtm. Discussed much with Potuz offline and this looks good to me.

**Pull request merged**